xds/internal/xdsclient/xdsresource/unmarshal_lds.go

```diff
@@ -152,6 +153,10 @@ func validateHTTPFilterConfig(cfg *anypb.Any, lds, optional bool) (httpfilter.Fi
 		}
 		return nil, nil, fmt.Errorf("no filter implementation found for %q", typeURL)
 	}
+	if rbac, ok := filterBuilder.(rbac.Builder); ok {
+		rbac.Name = name
```

Comment:
This looks very suspicious.  Why are we altering a specific in the builder of a particular type?

Reply:
Sorry for the delay in responding, I was out of office

So there's a reason it looks weird/messy - these filters have a name, but in the transition from v3httppb.HttpFilter to httpfilter.Filter we drop the Name. For audit logging this name is important and needs to be logged - [code](https://github.com/grpc/grpc-go/blob/642dd63a85275a96d172f446911fd04111e2c74c/internal/xds/rbac/rbac_engine.go#L295). Explanation with code links follows:

To step through the code:
v3httppb.HttpFilter [has a Name](https://github.com/grpc/grpc-go/blob/642dd63a85275a96d172f446911fd04111e2c74c/xds/internal/xdsclient/xdsresource/unmarshal_lds.go#L199).
However the internal httpfilter.Filter that we represent it as [does not have a name](https://github.com/grpc/grpc-go/blob/642dd63a85275a96d172f446911fd04111e2c74c/xds/internal/httpfilter/httpfilter.go#L38)
The RBAC proto from envoy [also doesn't have a name](https://github.com/envoyproxy/envoy/blob/6de02d23825523484941f1216c43c2b89f03752c/api/envoy/config/rbac/v3/rbac.proto#L80) and we use it directly [in rbac_engine.go to create our RBAC engine](https://github.com/grpc/grpc-go/blob/642dd63a85275a96d172f446911fd04111e2c74c/internal/xds/rbac/rbac_engine.go#L55)

In the non-xds case, this is [easy to pass through](https://github.com/grpc/grpc-go/blob/642dd63a85275a96d172f446911fd04111e2c74c/authz/grpc_authz_server_interceptors.go#L51).

This PR represents passing it through in the XDS case, which is more complicated and messy just because of the structure of the code explained above.
I'm open to other ways to do this, but we thought this was the least problematic.
The other options I see:
1. Add a name to httpfilter.Filter - this breaks a lot of code because it's a generic interface used elsewhere in the codebase. Also add the name in the RBAC proto, this is difficult because it requires a change to the proto definition in envoy.
1. Ignore the filter name in the XDS case and just have a mismatch for XDS vs. non-XDS audit logging, so the policy name will always be empty for XDS

Please let me know what you think - I'll focus on this conversation before addressing the other comments

Reply:
> 2. Ignore the filter name in the XDS case and just have a mismatch for XDS vs. non-XDS audit logging, so the policy name will always be empty for XDS

What's being done in the other languages here, logging-wise?  Are we showing the filter name or not?

> 1. Add a name to httpfilter.Filter - this breaks a lot of code because it's a generic interface used elsewhere in the codebase. Also add the name in the RBAC proto, this is difficult because it requires a change to the proto definition in envoy.

This seems to be the right way to go if we need to show the filter name.  I'm not sure exactly the best way to handle this - do we need `ParseFilterConfig` and `...Override` to accept another parameter for the filter's name?  Or do the interceptor builders accept it?  I'm thinking the latter, otherwise the API makes it possible to call the interceptor builders with configs & overrides that have mismatched names.

Both modifying a copy of the builder, and special casing this one filter in this general filter-handling code, both feel very wrong from a software engineering perspective.

Reply:
We are showing the filter name in other languages.

I agree that this feels wrong from a software eng perspective, I think changing `httpfilter.Filter` to have a name could be the correct thing to do because these RBACs can have names. However, `httpfilter.Filter` gets used outside of the RBAC setup/XDS case where we know the filters have names.
However, arguing against this - the other implementations don't have names. So I question if they should, or if maybe this Name on the RBAC filter is actually not important? Or, are these not correctly represented by the `httpfilter.Filter` abstraction?

You can see in the test code where exactly the name is coming from (it's on the `HttpFilter` that holds the RBAC proto).

Reply:
I'm confused by this:

> We are showing the filter name in other languages.
...
> the other implementations don't have names

Reply:
Ah, thanks for clarifying.

It looks like C++ does this (? - https://github.com/grpc/grpc/blob/d139c4a014b226af23c515a2e78e92600087e12c/src/core/ext/xds/xds_http_rbac_filter.cc#L584-L585), but Java does not.  If we do this, it should be in a general way, not just something hard-coded for the RBAC filter.

Reply:
Sorry for the delay, I was out Monday. I'm looking into what it would look like to do it more generally without breaking existing code